diffrax/_term.py

```diff
@@ -219,45 +219,55 @@ def _mul(v):
 """
 
 
-class _CallableToPath(AbstractPath):
+class _CallableToPath(AbstractPath[_Control]):
     fn: Callable
 
     @property
-    def t0(self):
+    def t0(self):  # pyright: ignore[reportIncompatibleVariableOverride]
         return -jnp.inf
 
     @property
-    def t1(self):
+    def t1(self):  # pyright: ignore[reportIncompatibleVariableOverride]
         return jnp.inf
 
     def evaluate(
         self, t0: RealScalarLike, t1: Optional[RealScalarLike] = None, left: bool = True
-    ) -> PyTree[Array]:
+    ) -> _Control:
         return self.fn(t0, t1)
 
 
-def _callable_to_path(x):
-    if isinstance(x, AbstractPath):
-        return x
-    else:
-        return _CallableToPath(x)
-
-
 # vf: Shaped[Array, "*state *control"]
 # control: Shaped[Array, "*control"]
 # return: Shaped[Array, "*state"]
 def _prod(vf, control):
     return jnp.tensordot(vf, control, axes=jnp.ndim(control))
 
 
-class _ControlTerm(AbstractTerm):
+class _AbstractControlTerm(AbstractTerm[_Control]):
     vector_field: Callable[[RealScalarLike, Y, Args], VF]
-    control: Union[AbstractPath, Callable] = eqx.field(converter=_callable_to_path)
+    _control: Union[
+        AbstractPath[_Control], Callable[[RealScalarLike, RealScalarLike], _Control]
+    ]
+
+    @property
+    def control(self) -> AbstractPath[_Control]:
+        return self._callable_to_path(self.control)
+
+    @staticmethod
+    def _callable_to_path(
+        x: Union[
+            AbstractPath[_Control], Callable[[RealScalarLike, RealScalarLike], _Control]
+        ],
+    ) -> AbstractPath[_Control]:
+        if isinstance(x, AbstractPath):
+            return x
+        else:
+            return _CallableToPath(x)
```

Comment:
To be honest, I'd probably sacrifice accuracy in the name of brevity here, and just do
```python
control: Union[AbstractPath[_Control], Callable[[RealScalarLike, RealScalarLike], _Control]] = eqx.field(converter=_callable_to_path)
```

This is a slightly-too-broad annotation for the field itself, of course, but I think it's more readable overall.